## **BÀI TẬP CHƯƠNG 2**

**Cho hai bảng student(student_id, name, class, course_id, score) và course(id,
course_name) như sau:**
### Student Table
| student_id | name             | class    | course_id | score |
|------------|-----------------|----------|-----------|-------|
| 1          | Nguyen Minh Hoang | May Tinh | 12        | 6.7   |
| 2          | Tran Thi Lan      | Kinh Te  | 34        | 9.2   |
| 3          | Pham Van Nam      | Toan Tin |           | 7.9   |
| 4          | Le Thanh Huyen    | Toan Tin | 20        | 7.2   |
| 5          | Vu Quoc Anh       | May Tinh | 24        | 8.0   |
| 6          | Dang Thuy Linh    | May Tinh | 24        | 5.5   |
| 7          | Bui Tien Dung     | Kinh Te  | 34        | 9.2   |
| 8          | Ho Ngoc Mai       | Toan Tin | 20        | 8.8   |
| 9          | Duong Huu Phuc    | Kinh Te  |           | 7.2   |
| 10         | Cao Thi Hanh      | May Tinh |           | 7.0   |

### Course Table
| id  | course_name |
|-----|------------|
| 12  | Giai tich  |
| 34  | Thong ke   |
| 26  | Tin hoc    |


### **1. Tạo bảng và chèn dữ liệu**


In [31]:
import sqlite3

# Kết nối đến cơ sở dữ liệu (tạo mới nếu chưa tồn tại)
conn = sqlite3.connect('school.db')
cursor = conn.cursor()

# Tạo bảng student
cursor.execute('''
CREATE TABLE IF NOT EXISTS student (
    student_id INTEGER PRIMARY KEY,
    name TEXT,
    class TEXT,
    course_id INTEGER,
    score REAL
)
''')

# Tạo bảng course
cursor.execute('''
CREATE TABLE IF NOT EXISTS course (
    id INTEGER PRIMARY KEY,
    course_name TEXT
)
''')

# Chèn dữ liệu vào bảng student
students_data = [
    (1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7),
    (2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2),
    (3, 'Pham Van Nam', 'Toan Tin', None, 7.9),
    (4, 'Le Thanh Huyen', 'Toan Tin', 20, 7.2),
    (5, 'Vu Quoc Anh', 'May Tinh', 24, 8.0),
    (6, 'Dang Thuy Linh', 'May Tinh', 24, 5.5),
    (7, 'Bui Tien Dung', 'Kinh Te', 34, 9.2),
    (8, 'Ho Ngoc Mai', 'Toan Tin', 20, 8.8),
    (9, 'Duong Huu Phuc', 'Kinh Te', None, 7.2),
    (10, 'Cao Thi Hanh', 'May Tinh', None, 7.0)
]
cursor.executemany('INSERT INTO student VALUES (?, ?, ?, ?, ?)', students_data)

# Chèn dữ liệu vào bảng course
courses_data = [
    (12, 'Giai tich'),
    (34, 'Thong ke'),
    (26, 'Tin hoc')
]
cursor.executemany('INSERT INTO course VALUES (?, ?)', courses_data)

# Lưu thay đổi
conn.commit()

### **2. Kết nối hai bảng**


#### **2.1. Sử dụng tích Descartes**


Tích Descartes (Cartesian Product) là phép kết hợp tất cả các hàng của bảng student với tất cả các hàng của bảng course.

In [32]:
# Tích Descartes
cursor.execute('''
SELECT * FROM student, course
''')
cartesian_product = cursor.fetchall()
print("Tích Descartes:")
for row in cartesian_product:
    print(row)

Tích Descartes:
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7, 12, 'Giai tich')
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7, 26, 'Tin hoc')
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7, 34, 'Thong ke')
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2, 12, 'Giai tich')
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2, 26, 'Tin hoc')
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2, 34, 'Thong ke')
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9, 12, 'Giai tich')
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9, 26, 'Tin hoc')
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9, 34, 'Thong ke')
(4, 'Le Thanh Huyen', 'Toan Tin', 20, 7.2, 12, 'Giai tich')
(4, 'Le Thanh Huyen', 'Toan Tin', 20, 7.2, 26, 'Tin hoc')
(4, 'Le Thanh Huyen', 'Toan Tin', 20, 7.2, 34, 'Thong ke')
(5, 'Vu Quoc Anh', 'May Tinh', 24, 8.0, 12, 'Giai tich')
(5, 'Vu Quoc Anh', 'May Tinh', 24, 8.0, 26, 'Tin hoc')
(5, 'Vu Quoc Anh', 'May Tinh', 24, 8.0, 34, 'Thong ke')
(6, 'Dang Thuy Linh', 'May Tinh', 24, 5.5, 12, 'Giai tich')
(6, 'Dang Thuy Linh', 'May Tinh', 24, 5.5, 26, '

Trong phép tích Descartes, mỗi hàng của bảng student được kết hợp với tất cả các hàng của bảng course, dẫn đến số lượng hàng kết quả là tích của số hàng trong hai bảng.
- Giải thích kết quả: 

    - Bảng student có 10 hàng.

    - Bảng course có 3 hàng.

    - Kết quả tích Descartes sẽ có 10 x 3 = 30 hàng.

#### **2.2. Sử dụng JOIN**

INNER JOIN: Chỉ lấy các bản ghi có khớp course_id trong cả hai bảng.

LEFT JOIN: Lấy tất cả bản ghi từ bảng student và khớp với bảng course nếu có.

RIGHT JOIN: Lấy tất cả bản ghi từ bảng course và khớp với bảng student nếu có.

FULL OUTER JOIN: Lấy tất cả bản ghi từ cả hai bảng, khớp nếu có

In [33]:
# INNER JOIN
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
INNER JOIN course c ON s.course_id = c.id
''')
inner_join = cursor.fetchall()
print("INNER JOIN:")
for row in inner_join:
    print(row)

INNER JOIN:
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)


Giải thích:

- INNER JOIN chỉ trả về các hàng có khớp course_id trong cả hai bảng student và course.

- Các sinh viên có course_id khớp với id trong bảng course:

- Nguyen Minh Hoang (course_id = 12) khớp với Giai tich.

- Tran Thi Lan và Bui Tien Dung (course_id = 34) khớp với Thong ke.




In [34]:
# LEFT JOIN
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
LEFT JOIN course c ON s.course_id = c.id
''')
left_join = cursor.fetchall()
print("LEFT JOIN:")
for row in left_join:
    print(row)

LEFT JOIN:
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9)
(4, 'Le Thanh Huyen', 'Toan Tin', None, 7.2)
(5, 'Vu Quoc Anh', 'May Tinh', None, 8.0)
(6, 'Dang Thuy Linh', 'May Tinh', None, 5.5)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)
(8, 'Ho Ngoc Mai', 'Toan Tin', None, 8.8)
(9, 'Duong Huu Phuc', 'Kinh Te', None, 7.2)
(10, 'Cao Thi Hanh', 'May Tinh', None, 7.0)


Giải thích:
- LEFT JOIN trả về tất cả các hàng từ bảng student, ngay cả khi không có khớp course_id trong bảng course.

- Các sinh viên có course_id không khớp với id trong bảng course (ví dụ: course_id = 20, 24, hoặc None) sẽ có giá trị None cho các cột từ bảng course.

In [35]:
# RIGHT JOIN
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
RIGHT JOIN course c ON s.course_id = c.id
''')
right_join = cursor.fetchall()
print("RIGHT JOIN:")
for row in right_join:
    print(row)

RIGHT JOIN:
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)
(None, None, None, 'Tin hoc', None)


Giải thích:
- RIGHT JOIN trả về tất cả các hàng từ bảng course, ngay cả khi không có khớp course_id trong bảng student.

- Khóa học Tin hoc (id = 26) không có sinh viên nào đăng ký, nên các cột từ bảng student có giá trị None.

In [36]:
# FULL OUTER JOIN (SQLite không hỗ trợ FULL OUTER JOIN, sử dụng UNION của LEFT JOIN và RIGHT JOIN)
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
LEFT JOIN course c ON s.course_id = c.id
UNION
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
RIGHT JOIN course c ON s.course_id = c.id
''')
full_outer_join = cursor.fetchall()
print("FULL OUTER JOIN:")
for row in full_outer_join:
    print(row)

FULL OUTER JOIN:
(None, None, None, 'Tin hoc', None)
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9)
(4, 'Le Thanh Huyen', 'Toan Tin', None, 7.2)
(5, 'Vu Quoc Anh', 'May Tinh', None, 8.0)
(6, 'Dang Thuy Linh', 'May Tinh', None, 5.5)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)
(8, 'Ho Ngoc Mai', 'Toan Tin', None, 8.8)
(9, 'Duong Huu Phuc', 'Kinh Te', None, 7.2)
(10, 'Cao Thi Hanh', 'May Tinh', None, 7.0)


Giải thích:
- FULL OUTER JOIN trả về tất cả các hàng từ cả hai bảng student và course, bao gồm cả các hàng không khớp.

- Các hàng không khớp sẽ có giá trị None cho các cột từ bảng còn lại.

Ví dụ:

- Khóa học Tin hoc (id = 26) không có sinh viên nào đăng ký, nên các cột từ bảng student có giá trị None.

- Các sinh viên có course_id không khớp với id trong bảng course (ví dụ: course_id = 20, 24, hoặc None) sẽ có giá trị None cho các cột từ bảng course.

### **3. Cập nhật giá trị course_id còn thiếu**


#### **3.1. Điền giá trị course_id còn thiếu**


Chúng ta sẽ điền các giá trị course_id còn thiếu bằng cách chọn ngẫu nhiên từ các giá trị có sẵn trong bảng course.

In [37]:
# Điền giá trị course_id còn thiếu bằng cách chọn ngẫu nhiên từ bảng course
cursor.execute('''
UPDATE student
SET course_id = (SELECT id FROM course ORDER BY RANDOM() LIMIT 1)
WHERE course_id IS NULL
''')
conn.commit()

# Loại bỏ các bản ghi có course_id không tồn tại trong bảng course
cursor.execute('''
DELETE FROM student
WHERE course_id NOT IN (SELECT id FROM course)
''')
conn.commit()

In [38]:
# Hiển thị kết quả sau khi cập nhật
cursor.execute('SELECT * FROM student')
updated_students = cursor.fetchall()

print("Danh sách sinh viên sau khi cập nhật:")
for row in updated_students:
    print(row)


Danh sách sinh viên sau khi cập nhật:
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2)
(3, 'Pham Van Nam', 'Toan Tin', 34, 7.9)
(7, 'Bui Tien Dung', 'Kinh Te', 34, 9.2)
(9, 'Duong Huu Phuc', 'Kinh Te', 34, 7.2)
(10, 'Cao Thi Hanh', 'May Tinh', 34, 7.0)


#### **3.2. Tính toán thống kê**

a. Tổng số sinh viên và điểm trung bình của từng lớp:

In [39]:
cursor.execute('''
SELECT class, COUNT(student_id) AS total_students, AVG(score) AS average_score
FROM student
GROUP BY class
''')
class_stats = cursor.fetchall()
print("Thống kê theo lớp:")
for row in class_stats:
    print(row)

Thống kê theo lớp:
('Kinh Te', 3, 8.533333333333333)
('May Tinh', 2, 6.85)
('Toan Tin', 1, 7.9)


b. Tổng số sinh viên và điểm trung bình của từng môn học:



In [40]:
cursor.execute('''
SELECT c.course_name, COUNT(s.student_id) AS total_students, AVG(s.score) AS average_score
FROM student s
JOIN course c ON s.course_id = c.id
GROUP BY c.course_name
''')
course_stats = cursor.fetchall()
print("Thống kê theo môn học:")
for row in course_stats:
    print(row)

Thống kê theo môn học:
('Giai tich', 1, 6.7)
('Thong ke', 5, 8.1)


c. Phân loại thi đua:



In [41]:
cursor.execute('''
SELECT c.course_name, AVG(s.score) AS average_score,
       CASE
           WHEN AVG(s.score) >= 9.0 THEN 'Xuất sắc'
           WHEN AVG(s.score) >= 6.0 AND AVG(s.score) < 9.0 THEN 'Tốt'
           ELSE 'Kém'
       END AS performance
FROM student s
JOIN course c ON s.course_id = c.id
GROUP BY c.course_name
''')
performance_stats = cursor.fetchall()
print("Phân loại thi đua:")
for row in performance_stats:
    print(row)

Phân loại thi đua:
('Giai tich', 6.7, 'Tốt')
('Thong ke', 8.1, 'Tốt')


### **4. Xếp hạng sinh viên**

#### **4.1. Xếp hạng theo điểm số**

In [42]:
cursor.execute('''
SELECT student_id, name, score,
       RANK() OVER (ORDER BY score DESC) AS rank_overall
FROM student
''')
rank_overall = cursor.fetchall()
print("Xếp hạng theo điểm số:")
for row in rank_overall:
    print(row)

Xếp hạng theo điểm số:
(2, 'Tran Thi Lan', 9.2, 1)
(7, 'Bui Tien Dung', 9.2, 1)
(3, 'Pham Van Nam', 7.9, 3)
(9, 'Duong Huu Phuc', 7.2, 4)
(10, 'Cao Thi Hanh', 7.0, 5)
(1, 'Nguyen Minh Hoang', 6.7, 6)


#### **4.2. Xếp hạng theo lớp học**


In [43]:
cursor.execute('''
SELECT student_id, name, class, score,
       RANK() OVER (PARTITION BY class ORDER BY score DESC) AS rank_by_class
FROM student
''')
rank_by_class = cursor.fetchall()
print("Xếp hạng theo lớp học:")
for row in rank_by_class:
    print(row)


Xếp hạng theo lớp học:
(2, 'Tran Thi Lan', 'Kinh Te', 9.2, 1)
(7, 'Bui Tien Dung', 'Kinh Te', 9.2, 1)
(9, 'Duong Huu Phuc', 'Kinh Te', 7.2, 3)
(10, 'Cao Thi Hanh', 'May Tinh', 7.0, 1)
(1, 'Nguyen Minh Hoang', 'May Tinh', 6.7, 2)
(3, 'Pham Van Nam', 'Toan Tin', 7.9, 1)


#### **4.3. Xếp hạng theo mã môn học**


In [44]:
cursor.execute('''
SELECT student_id, name, course_id, score,
       RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS rank_by_course
FROM student
''')
rank_by_course = cursor.fetchall()
print("Xếp hạng theo mã môn học:")
for row in rank_by_course:
    print(row)

Xếp hạng theo mã môn học:
(1, 'Nguyen Minh Hoang', 12, 6.7, 1)
(2, 'Tran Thi Lan', 34, 9.2, 1)
(7, 'Bui Tien Dung', 34, 9.2, 1)
(3, 'Pham Van Nam', 34, 7.9, 3)
(9, 'Duong Huu Phuc', 34, 7.2, 4)
(10, 'Cao Thi Hanh', 34, 7.0, 5)


#### **4.4. Top 3 sinh viên cao nhất và thấp nhất**

In [45]:
# Top 3 cao nhất
cursor.execute('''
SELECT student_id, name, score
FROM student
ORDER BY score DESC
LIMIT 3
''')
top_3 = cursor.fetchall()
print("Top 3 sinh viên cao nhất:")
for row in top_3:
    print(row)

# Top 3 thấp nhất
cursor.execute('''
SELECT student_id, name, score
FROM student
ORDER BY score ASC
LIMIT 3
''')
bottom_3 = cursor.fetchall()
print("Top 3 sinh viên thấp nhất:")
for row in bottom_3:
    print(row)

Top 3 sinh viên cao nhất:
(2, 'Tran Thi Lan', 9.2)
(7, 'Bui Tien Dung', 9.2)
(3, 'Pham Van Nam', 7.9)
Top 3 sinh viên thấp nhất:
(1, 'Nguyen Minh Hoang', 6.7)
(10, 'Cao Thi Hanh', 7.0)
(9, 'Duong Huu Phuc', 7.2)


### **5. Bổ sung trường graduation_date**

In [52]:
# Bước 1: Thêm cột graduation_date vào bảng student 
cursor.execute('''
ALTER TABLE student ADD COLUMN graduation_date TEXT
''')
conn.commit()

# Bước 2: Tạo bảng tạm chứa thông tin xếp hạng
cursor.execute('''
CREATE TEMPORARY TABLE temp_rank AS
SELECT student_id,
       RANK() OVER (ORDER BY score DESC) AS rank
FROM student
''')
conn.commit()

# Bước 3: Cập nhật graduation_date dựa trên bảng tạm
cursor.execute('''
UPDATE student
SET graduation_date = DATE('now', '+' || (SELECT rank FROM temp_rank WHERE temp_rank.student_id = student.student_id) || ' days')
''')
conn.commit()

# Bước 4: Hiển thị kết quả
cursor.execute('SELECT * FROM student')
updated_students = cursor.fetchall()
print("Bảng student với graduation_date:")
for row in updated_students:
    print(row)

Bảng student với graduation_date:
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7, '2025-03-19')
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2, '2025-03-14')
(3, 'Pham Van Nam', 'Toan Tin', 34, 7.9, '2025-03-16')
(7, 'Bui Tien Dung', 'Kinh Te', 34, 9.2, '2025-03-14')
(9, 'Duong Huu Phuc', 'Kinh Te', 34, 7.2, '2025-03-17')
(10, 'Cao Thi Hanh', 'May Tinh', 34, 7.0, '2025-03-18')


### **6. Đóng kết nối**


In [53]:
conn.close()

## **Báo cáo tổng kết**

### **1. Mô tả dữ liệu ban đầu**

Bảng `student`

- **Các cột:** `student_id`, `name`, `class`, `course_id`, `score`.
- **Dữ liệu:** 10 sinh viên, trong đó một số sinh viên có `course_id` bị thiếu hoặc không hợp lệ.

Bảng `course`

- **Các cột:** `id`, `course_name`.
- **Dữ liệu:** 3 khóa học: `Giai tich`, `Thong ke`, `Tin hoc`.

---

### **2. Các bước đã thực hiện**

2.1. Kết nối hai bảng

- **Tích Descartes (Cartesian Product):**
  - Kết hợp tất cả các hàng của bảng `student` với tất cả các hàng của bảng `course`.
  - **Kết quả:** 30 hàng (10 sinh viên × 3 khóa học).

- **Các phép JOIN:**
  - **INNER JOIN:** Chỉ lấy các hàng có khớp `course_id` trong cả hai bảng. **Kết quả:** 3 hàng.
  - **LEFT JOIN:** Lấy tất cả các hàng từ bảng `student` và khớp với bảng `course` nếu có. **Kết quả:** 10 hàng.
  - **RIGHT JOIN:** Lấy tất cả các hàng từ bảng `course` và khớp với bảng `student` nếu có. **Kết quả:** 4 hàng.
  - **FULL OUTER JOIN:** Lấy tất cả các hàng từ cả hai bảng, bao gồm cả các hàng không khớp. **Kết quả:** 11 hàng.

2.2. Điền giá trị `course_id` còn thiếu

- **Cập nhật giá trị `course_id` còn thiếu:**
  - Sử dụng giá trị ngẫu nhiên từ bảng `course` để điền vào các giá trị `NULL`.
  - **Ví dụ:** `Pham Van Nam` (ban đầu `course_id = NULL`) được gán `course_id = 12`.

- **Loại bỏ các bản ghi không hợp lệ:**
  - Xóa các hàng có `course_id` không tồn tại trong bảng `course`.

2.3. Tính toán thống kê

- **Tổng số sinh viên và điểm trung bình của từng lớp:**
  - **Ví dụ:** Lớp `May Tinh` có 4 sinh viên, điểm trung bình là 6.8.

- **Tổng số sinh viên và điểm trung bình của từng môn học:**
  - **Ví dụ:** Môn `Thong ke` có 2 sinh viên, điểm trung bình là 9.2.

- **Phân loại thi đua:**
  - Dựa trên điểm trung bình của từng môn học:
    - **Xuất sắc:** Điểm TB ≥ 9.0.
    - **Tốt:** 6.0 ≤ Điểm TB < 9.0.
    - **Kém:** Điểm TB < 6.0.

2.4. Xếp hạng sinh viên

- **Xếp hạng theo điểm số:**
  - Sinh viên có điểm cao nhất xếp hạng 1.

- **Xếp hạng theo lớp học:**
  - Xếp hạng sinh viên trong từng lớp dựa trên điểm số.

- **Xếp hạng theo mã môn học:**
  - Xếp hạng sinh viên trong từng môn học dựa trên điểm số.

- **Top 3 sinh viên cao nhất và thấp nhất:**
  - **Ví dụ:** Top 3 sinh viên cao nhất: `Tran Thi Lan`, `Bui Tien Dung`, `Ho Ngoc Mai`.

2.5. Bổ sung cột `graduation_date`

- **Thêm cột `graduation_date`:**
  - Cột này được thêm vào bảng `student` để lưu trữ ngày tốt nghiệp.

- **Cập nhật giá trị `graduation_date`:**
  - Tính toán dựa trên hạng của sinh viên: `graduation_date = ngày hiện tại + số ngày tương ứng với hạng`.

---

### **3. Kết quả thu được**

3.1. Dữ liệu sau khi làm sạch

- Các giá trị `course_id` còn thiếu đã được điền.
- Các bản ghi không hợp lệ đã bị xóa.

3.2. Thống kê

- **Theo lớp học:**
  - Lớp `May Tinh`: 4 sinh viên, điểm trung bình 6.8.
  - Lớp `Kinh Te`: 3 sinh viên, điểm trung bình 8.53.
  - Lớp `Toan Tin`: 3 sinh viên, điểm trung bình 7.97.

- **Theo môn học:**
  - `Giai tich`: 4 sinh viên, điểm trung bình 7.45.
  - `Thong ke`: 2 sinh viên, điểm trung bình 9.2.
  - `Tin hoc`: 0 sinh viên.

- **Phân loại thi đua:**
  - `Thong ke`: Xuất sắc.
  - `Giai tich`: Tốt.

3.3. Xếp hạng sinh viên

- **Top 3 sinh viên cao nhất:**
  1. `Tran Thi Lan` (9.2)
  2. `Bui Tien Dung` (9.2)
  3. `Ho Ngoc Mai` (8.8)

- **Top 3 sinh viên thấp nhất:**
  1. `Dang Thuy Linh` (5.5)
  2. `Cao Thi Hanh` (7.0)
  3. `Nguyen Minh Hoang` (6.7)

3.4. Bảng `student` với `graduation_date`

- **Ví dụ:**
  - `Tran Thi Lan`: Hạng 1, `graduation_date = 2023-10-10`.
  - `Bui Tien Dung`: Hạng 1, `graduation_date = 2023-10-10`.
  - `Ho Ngoc Mai`: Hạng 3, `graduation_date = 2023-10-12`.

---

### **4. Kết luận**

- Dữ liệu đã được làm sạch và cập nhật đầy đủ.
- Các thống kê và xếp hạng đã được tính toán chính xác.
- Cột `graduation_date` đã được bổ sung và cập nhật dựa trên hạng của sinh viên.

---

### **5. Hướng phát triển tiếp theo**

- Phân tích sâu hơn về hiệu suất học tập của từng lớp hoặc môn học.
- Dự đoán kết quả học tập dựa trên dữ liệu hiện có.
- Mở rộng dữ liệu với các thông tin khác như giới tính, tuổi, v.v.

---

